base/abstractarray.jl

```diff
@@ -1952,6 +1911,7 @@ function typed_hvcat(::Type{T}, rows::Tuple{Vararg{Int}}, xs::Number...) where T
 end
 
 function typed_hvcat(::Type{T}, rows::Tuple{Vararg{Int}}, as...) where T
+    all(x -> isa(x, Number) || isa(x, AbstractArray{<:Number}), as) && return typed_hvcat1(T, rows, as...)
```

Comment:
I don't think these need to be different functions, I think you can just introduce a new function `_size` which returns `size` for `AbstractArrays` and `()` otherwise.

Reply:
Would that break a type that created its own typed_hcat and typed_vcat but not typed_hvcat?

Reply:
Since the new signature is strictly wider, I don't think this should cause any issues.

Reply:
Another issue is that method only handles vectors and 2d arrays, still need the fallback for larger; later maybe could be merged with my hvncat for a more general solution.